### ISSUE TITLE
E2E: Add CSI snapshot tests to kind cluster

### ISSUE THREAD

**Issue by kaovilai** [velero-io/velero#7507]
Please do not start working on this issue 

**Describe the problem/challenge you have**

I have found some examples of adding CSI snapshots in KinD.

https://github.com/topolvm/topolvm/tree/main/example
https://github.com/topolvm/topolvm/blob/main/test/e2e/Makefile#L267

and that they could be run inside github actions.

Maybe https://github.com/kubernetes-csi/csi-driver-host-path can also be used
https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/docs/example-snapshot-metadata.md to test https://github.com/kubernetes/enhancements/issues/3314

Supports testing volumeGroupSnapshot
https://github.com/kubernetes-csi/csi-driver-host-path/blob/8b88d04ec388312df1b910bf1cb48926c63c2660/pkg/hostpath/groupcontrollerserver.go#L45

**Describe the solution you'd like**

It would be good to enable CSI tests in PR tests.

**Anything else you would like to add:**

/backlog

**Environment:**

- Velero version (use `velero version`):
- Kubernetes version (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "The project would be better with this feature added"
- :-1: for "This feature will not enhance the project in a meaningful way"

**Comment by kaovilai** [comment]
Unstale

**Comment by kaovilai** [comment]
unstale

**Comment by kaovilai** [comment]
unstale

**Comment by kaovilai** [comment]
https://github.com/kubernetes-csi/csi-driver-host-path is used by minikube so perhaps we can follow.

https://minikube.sigs.k8s.io/docs/tutorials/volume_snapshots_and_csi/

**Comment by kaovilai** [comment]
@mpryc Thoughts if you can pick this up?

**Comment by kaovilai** [comment]
cc: @shubham-pampattiwar volumeGroupSnapshot is also supported by this driver https://github.com/kubernetes-csi/csi-driver-host-path/blob/8b88d04ec388312df1b910bf1cb48926c63c2660/pkg/hostpath/groupcontrollerserver.go#L45

**Comment by kaovilai** [comment]
We won't get to this in 1.18.

**Comment by kadeksuryam** [comment]
Hi @kaovilai  and @shubham-pampattiwar ,

I'm planning to apply for the 2026 Term 3 LFX mentorship project linked to this issue.

I'm still getting familiar with Velero. Right now, I'm learning about its backup, restore, and CSI snapshot flow. After that, I plan to look through the existing E2E tests and run the current Kind setup locally.

I've worked with CI pipelines before, and from my initial reading, I see that Velero relies heavily on CRDs and controllers. I've built a Kubernetes controller in Go, so those patterns are familiar to me. I also noticed the note asking people not to start working on this issue, so I'll keep this to local exploration for now.

If there's a particular part of Velero or the test setup that you think applicants should understand first, I'd appreciate the recommendation.

Thanks,
Surya

**Comment by true-brace05** [comment]
Hi @kaovilai  and @shubham-pampattiwar 
I'm planning to apply for the LFX Mentorship project (2026 Term 3 ) associated with this issue, and I've started reproducing the Kind environment locally to better understand the current gap before making any implementation attempts.
So far, I've:
Set up the existing Velero Kind E2E environment locally.
Traced the current E2E workflow (.github/workflows/e2e-kind.yml) to understand where the Kind cluster is created and where the tests start.
Created a local Kind cluster and confirmed that it doesn't include the VolumeSnapshot CRDs by default.
Tried deploying csi-driver-host-path using its deploy.sh script (one of the references mentioned in this issue).
During that setup, the deployment failed while creating the VolumeSnapshotClass with:
no matches for kind "VolumeSnapshotClass" in version "snapshot.storage.k8s.io/v1"
ensure CRDs are installed first
From what I understand, this happens because a fresh Kind cluster doesn't include the snapshot CRDs and Snapshot Controller, which matches the Minikube documentation where the volumesnapshots addon installs those components before the CSI HostPath driver.
Before I continue investigating, I wanted to ask one design question:
For this issue, is the expected direction to install the snapshot CRDs and Snapshot Controller (similar to the Minikube/TopoLVM setup) before deploying csi-driver-host-path, or is there another preferred approach for the Kind E2E environment?
I'm continuing to explore the existing E2E tests locally, but I'd like to make sure I'm investigating in the intended direction before going further.
Thanks!

**Comment by harshitsaini17** [comment]
I looked into why the CSI specs never run on the Kind pipeline, to understand
the scope here. Per the note at the top, this is investigation only, I haven't
started on any implementation.

There are two independent gaps, and closing either one alone still executes
zero CSI specs:

**1. The Ginkgo matrix is an allowlist, and CSI has no entry.**
`.github/workflows/e2e-test-kind.yaml` runs exactly four label expressions:

    "Basic && (ClusterResource || NodePort || StorageClass)"
    "ResourceFiltering && !FSBackup"
    "ResourceModifier || (Backups && BackupsSync) || PrivilegesMgmt || OrderedResources"
    "(NamespaceMapping && Single && FSBackup) || (NamespaceMapping && Multiple && FSBackup)"

There's no `!Snapshot` exclusion anywhere, the `Snapshot` / `CSISnapshot` /
`CSIDataMover` specs are simply never selected, so even with a full CSI stack
installed on the cluster, nothing changes until a fifth matrix entry is added.
Related: the workflow doesn't set `FEATURES` at all, so `EnableCSI` stays off
regardless of which labels run.

**2. Nothing installs the snapshot stack on the Kind cluster.** No file in
`test/` or `.github/` references csi-driver-host-path, the snapshot CRDs, or
the snapshot-controller; that step is entirely unbuilt. The Go side is
already prepared, though: `test/util/csi/common.go` uses the
external-snapshotter v8 API types and clientset. And #10236 covers the missing
`testdata/volume-snapshot-class/kind.yaml` that `BeforeSuite` fails on first
once CSI is enabled.

One design question for whenever this gets picked up: csi-driver-host-path's
own deploy manifests create `csi-hostpath-snapclass`, which carries only
`app.kubernetes.io/*` labels and not the `velero.io/csi-volumesnapshot-class`
label Velero uses to select a VolumeSnapshotClass. Would the preferred approach
be to patch that label onto the driver-provided class during install, or to
apply a Velero-owned VolumeSnapshotClass for the hostpath driver, the way the
existing per-provider testdata files do?

Happy to write up the full design (install step + matrix entry + label
expression) if that would be useful.

**Comment by Shivampal157** [comment]
Hi @kaovilai @shubham-pampattiwar

I'm planning to apply for the LFX Term 3 project tied to this issue. I saw the note not to start work on it, so I'm only looking around locally for now.

From what I can tell so far, CSI tests already exist in the suite, but Kind CI never really sets them up. The Kind workflow doesn't install snapshot CRDs / snapshot-controller / host-path, and the Ginkgo labels for Snapshot / CSISnapshot / CSIDataMover don't seem to get picked in the matrix either. So even if the driver was there, those specs still wouldn't run.

I forked the repo (https://github.com/Shivampal157/velero) and I'm going through the Kind e2e workflow + the existing CSI test helpers next. One thing I'm still unsure about: for Kind, do you prefer installing external-snapshotter first and then csi-driver-host-path (like minikube), and for the VolumeSnapshotClass should we reuse the driver's class with the Velero label or add a kind-specific one under testdata?

Happy to share notes once I've tried the Kind setup myself. Not asking to be assigned.

**Comment by shubham-pampattiwar** [comment]
Thanks for the strong interest, everyone. This issue is tied to a CNCF LFX Mentorship 2026 Term 3 project, so I want to set clear expectations to keep things fair for all applicants.

**Please hold off on implementation PRs** for the core work here (CSI driver setup in Kind CI, enabling the existing CSI tests, and adding VolumeGroupSnapshot coverage). This is the mentorship project itself, and we want to keep it available for the selected mentee.

What you're welcome to do in the meantime:
- Reproduce the Kind environment locally to get familiar with the setup
- Read through the existing CSI test files and the e2e workflow
- Ask questions here about scope or the codebase

**How to apply:** Mentee applications open Aug 3 and close Aug 18 on the LFX Mentorship platform. Apply there once the project is listed. Selection happens through the LFX process.

Appreciate the enthusiasm, and looking forward to your applications.

**Comment by harshitsaini17** [comment]
Thanks to @shubham-pampattiwar for the suggestion. I got the stack running locally on kind. I ran into a couple of issues that aren't obvious from the manifests.

I set up kind on k8s v1.36.1, external-snapshotter v8.6.0, and csi-driver-host-path at `ffe5cb0`, deployed with its own `deploy/kubernetes-1.34/deploy.sh`.

The VGS part cost me some time. VolumeGroupSnapshots do nothing unless `--feature-gates=CSIVolumeGroupSnapshot=true` is set on both the snapshot-controller and the `csi-snapshotter` sidecar. Neither project's deploy manifests include this setting. The API server accepts the object fine, but it results in an empty `status`, no content, and no events. I thought I had misconfigured something until I found the flag in the external-snapshotter README. Once I enabled it on both parts, the same manifest showed `readyToUse: true` in about 25 seconds.

The second issue is with `e2e-storage-class` in `test/testdata/storage-class/kind.yaml`. It uses `rancher.io/local-path`, and the CSI specs build their PVCs from that same name (`test/types.go:29` → `backup-volume-info/base.go:121`). Local-path volumes aren't CSI volumes, so I wouldn't expect anything snapshottable there, even with the VolumeSnapshotClass from #10236 in place. I haven't run the specs to confirm this. The hostpath driver doesn't provide a StorageClass either. Once I added one with `provisioner: hostpath.csi.k8s.io`, a plain VolumeSnapshot worked end to end, showing `readyToUse: true` and 64Mi.

This raises a question I can't answer well: would you prefer to change `e2e-storage-class` to point at the hostpath driver so everything shares one class, or to add a CSI-only class that is applied when CSI is enabled? The first option alters the storage backing for specs that currently use local-path. The second option would add another `BeforeSuite` condition next to the one for VolumeSnapshotClass.

(My earlier question about the VolumeSnapshotClass labeling is no longer relevant as #10236 covers it.)

**Comment by prashant-zo** [comment]
Hi @shubham-pampattiwar @kaovilai

Following the note here I kept it to local exploration only, no implementation.

To understand the gap I created a fresh Kind cluster locally and confirmed it ships without the snapshot stack:

kubectl get crd | grep -i snapshot -> No resources found
kubectl get pods -A | grep -i snapshot -> nothing

So a default Kind cluster has neither the VolumeSnapshot CRDs nor the snapshot-controller, which lines up with what others have already noted here.

I also went through .github/workflows/e2e-test-kind.yaml and the test Makefile. From what I can see there are two separate gaps: the Kind workflow never installs external-snapshotter CRDs, the snapshot controller or csi-driver-host-path, and the Ginkgo label matrix only selects four label expressions, none of which include Snapshot, CSISnapshot or CSIDataMover. So even with the CSI stack installed, those specs would still be filtered out until a matrix entry is added. I also saw that test/util/csi/common.go already uses the external-snapshotter v8 client, so the Go side is prepared.

My background is in Kubernetes platform and CI work (Kind, GitHub Actions, and backup/restore with WAL archive and PITR), so this project fits well with what I want to learn. Happy to hear if there is a particular part of the setup you would suggest applicants understand first.

Thanks.

**Comment by kaovilai** [comment]
> the Ginkgo label matrix only selects four label expressions, none of which include Snapshot, CSISnapshot or CSIDataMover.

That has been intentional so yes by the end of this, we would hope those are run on KiND and filters are modified :)

> Happy to hear if there is a particular part of the setup you would suggest applicants understand first.

Looks like you already covered most of it.

**Comment by Pasta-coder** [comment]
I applied for this project . 

thankyou .

**Comment by Hamid0004** [comment]
Hi @kaovilai @shubham-pampattiwar,
I have submitted my application for this project on the LFX Mentorship portal. 

Following your guidance, I am keeping my focus strictly on local exploration, amiliarizing myself with the existing Ginkgo test suite, the `external-snapshotter` / `csi-driver-host-path` components, and the KinD CI workflow gaps.

Looking forward to the review process. Thanks!

**Comment by PranjalManhgaye** [comment]
Hi @shubham-pampattiwar @kaovilai,

I've been going through the comments here and poking at #10274 on my machine, not opening anything, just trying to understand the shape of the work for the LFX project.

Most of the basics are already covered better than I could say them (missing CRDs on Kind, install order, VGS needing the feature gate on both controllers. So I'll skip all that.

What I'm still stuck on is how #10274 fits with the existing Kind jobs. It adds a separate CSI workflow, which makes sense for CI time, but it also flips kind.yaml to `hostpath.csi.k8s.io` for everyone. The four current matrix jobs don't run the install script, so they'd still point PVCs at a StorageClass with no driver behind it. @harshitsaini17 already asked local-path vs hostpath, I'm wondering if the plan is a `CSI-only StorageClass` when `FEATURES=EnableCSI`, or eventually installing the snapshot stack in every Kind run.

Something else I only noticed from #10274: it pins `snapshotter v8.6.0` and waits until `csi-snapshotter reports all 4 caches populated`. I hadn't seen that mentioned in the thread, but it explains weird failures, if the sidecar is watching group snapshot APIs the CRDs don't actually serve, snapshots just never go ready and there's nothing useful in events. Felt like a Velero bug until I read the PR diff.

On scope, there's still no VGS test under` test/e2e/` as far as I can tell. Velero's docs say `1.18.1+` wants `external-snapshotter v8.2.0+ for v1beta2`, so if group snapshots are part of this issue, I'm curious whether you'd want a new describe block or something in the backup-volume-info pattern. Same question for VolumeGroupSnapshotClass, patch what host-path ships, or add something under testdata like #10236 did for VolumeSnapshotClass?

The issue also mentions KEP-3314 and host-path's snapshot-metadata docs. That seems more tied to the block mover stuff later,  is this mentorship basically `"get VolumeSnapshot + VGS green on Kind"` and leave `metadata/CBT` for another day?

Been using the `KEPs (1900, 3476, 3314)`, the `external-snapshotter README`, `Velero's VGS doc`, and the `minikube snapshot tutorial @kaovilai linked earlier`. Happy to jot down install pins / matrix sizing in a short doc if that's useful for whoever gets picked.

Thanks!

**Comment by shubham-pampattiwar** [comment]
This is being worked on as an LFX Mentorship project (CNCF Velero, 2026 Term 3).

Mentee: Krishna Tupe (@krishhna24)
Mentors: @shubham-pampattiwar, @kaovilai

Below is the plan and a checklist we'll keep updated as work lands. Feedback welcome.

## Goal

Install CSI snapshot infrastructure in Velero's Kind-based e2e CI, enable the existing CSI/snapshot tests that don't run there today, add VolumeGroupSnapshot (VGS) e2e coverage, and document it.

## Current state (verified against `main`)

- CI workflow `.github/workflows/e2e-test-kind.yaml`: the `setup-test-matrix` job runs 4 hardcoded Ginkgo label filters, none of which include snapshot or CSI:
  - `Basic && (ClusterResource || NodePort || StorageClass)`
  - `ResourceFiltering && !FSBackup`
  - `ResourceModifier || (Backups && BackupsSync) || PrivilegesMgmt || OrderedResources`
  - `(NamespaceMapping && Single && FSBackup) || (NamespaceMapping && Multiple && FSBackup)`
- CI installs only kind + MinIO. No external-snapshotter, no snapshot controller, no csi-driver-host-path. The CSI infra install is net-new.
- CSI/snapshot tests already exist but are not run in Kind CI:
  - `test/e2e/basic/backup-volume-info/csi_snapshot.go` (labels: `BackupVolumeInfo`, `CSISnapshot`)
  - `test/e2e/basic/backup-volume-info/csi_data_mover.go` (labels: `BackupVolumeInfo`, `CSIDataMover`)
  - Many `Snapshot`-labeled tests in `test/e2e/e2e_suite_test.go` (backup deletion, TTL, BSL deletion, migration, namespace-mapping, upgrade), most tagged `SkipVanillaZfs` (need a snapshot-capable provider).
- No VolumeGroupSnapshot/VGS tests exist yet. Fully net-new.
- Test runner: `make -C test/ run-e2e` passes `GINKGO_LABELS` to `--label-filter` (`test/Makefile`). The CSI path uses `FEATURES=EnableCSI` plus `PLUGINS=<provider-plugin-images>` (see `test/e2e/README.md`).

## Plan

### 1. CSI infra in Kind

- [ ] Install external-snapshotter (snapshot CRDs + controller) and csi-driver-host-path into a Kind cluster, documented step by step.
- [ ] Script that setup so it is reproducible and CI can reuse it.

Done when: a Kind cluster can provision `VolumeSnapshot`s, driven by a script.

### 2. Enable existing CSI/snapshot tests

- [ ] Run the existing `CSISnapshot` and `CSIDataMover` tests in `test/e2e/basic/backup-volume-info/` locally against the CSI-enabled cluster (`FEATURES=EnableCSI` + AWS provider plugin, used to access MinIO S3-compatible storage); triage and fix failures.
- [ ] Assess which `Snapshot`/`SkipVanillaZfs` tests can run on hostpath CSI.

### 3. Wire into CI

- [ ] Add a CSI-enabled label filter to the matrix in `setup-test-matrix` (e.g. a `CSISnapshot || CSIDataMover` entry).
- [ ] Add the CSI infra install steps to the `run-e2e-test` job.
- [ ] Stabilize flakes, timeouts, and ordering.

Done when: CSI/snapshot e2e tests pass in CI on PRs.

### 4. VolumeGroupSnapshot coverage

- [ ] Add net-new VGS e2e scenarios (backup/restore of a group of PVCs).
- [ ] Give them a `VolumeGroupSnapshot`/`VGS` label; run locally, then in CI.

Done when: VGS e2e scenarios are merged and running in CI.

### 5. Docs and wrap-up

- [ ] Update `test/e2e/README.md` with the CSI setup and how to add/run CSI e2e tests from Kind locally.
  - [ ] Bonus: a Makefile target to run the e2e suite that sets up Kind + MinIO and tears them down without touching existing local dev kubeconfigs (kind kubeconfig written to a temp location), so others can test the project during development.
- [ ] Write a blog post for the Velero/CNCF blog.
- [ ] Land remaining PRs and close loose ends.

## Related work we're building on

- #10236 (merged): added the kind `VolumeSnapshotClass` test data.
- #10267 (open): docs for the local Kind setup for CSI e2e.

**Comment by krishhna24** [comment]
Thanks for writing this up, and for the mentorship. The plan matches what I've been finding. Below is where I've got to against phases 1 and 2, plus three things I think are worth folding into the checklist.

## Phases 1 and 2

The stack is running on kind and the two existing CSI specs pass:

```
FEATURES=EnableCSI GINKGO_LABELS='BackupVolumeInfo && (CSISnapshot || CSIDataMover)'
Ran 2 of 68 Specs in 314.416 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 66 Skipped
```

`csi-snapshot-volumeinfo` at 217s and `csi-data-mover-volumeinfo` at 75s, against external-snapshotter v8.6.0, csi-driver-host-path v1.18.0, MinIO, and a locally built `velero:pr-test-linux-amd64`.

One caveat I want to be explicit about. That run needed `test/testdata/storage-class/kind.yaml` temporarily pointed at `hostpath.csi.k8s.io`, because `test/types.go:29` hardcodes `e2e-storage-class` and `InstallStorageClasses` (`test/util/velero/velero_utils.go:1510`) applies that file directly. So it demonstrates that the CSI half works, not that the global flip is acceptable. I reverted it.

That flip is also what reddens the twelve existing jobs on #10274, and the mechanism is narrow: the install step is added only inside the new `run-csi-e2e-test` job, so the pre-existing jobs get a StorageClass with no driver behind it. Only `storage-class-changing.go` provisions a PVC, and it spends 910s waiting on a PVC that can never bind before failing. Same label expression on a clean tree is 5 of 68 in 201s, all passing. So the StorageClass strategy looks like the real design decision in phase 3, and #8765 reads like its principled answer rather than a global flip.

Separately, both `run-csi-e2e-test` jobs on #10274 died at about 53s with `./hack/install-csi-hostpath.sh: Permission denied`, exit 126. The file is mode 100644. Worth knowing that the installer there has never actually executed in CI, so its red checks say nothing about the approach.

## VolumeGroupSnapshot works on kind

Verified again today on kind node image v1.36.1, external-snapshotter v8.6.0 for both CRDs and controller, csi-snapshotter v8.6.0 and hostpathplugin v1.17.1 as deployed by the driver's `kubernetes-latest` manifests. Two PVCs behind a label selector, one VolumeGroupSnapshot:

```
probe-vgs   readyToUse=true   groupsnapcontent-37865dd8-...   8s
member VolumeSnapshots: both readyToUse=true, 64Mi each
VGSContent: readyToUse=true, driver hostpath.csi.k8s.io, policy Delete
```

Two details that cost people time earlier in this thread. The `CSIVolumeGroupSnapshot=true` gate is needed on both the snapshot-controller and the `csi-snapshotter` sidecar, and neither project's deploy manifests set it. And the group snapshot CRDs sit under `groupsnapshot.storage.k8s.io`, not `snapshot.storage.k8s.io`, so the obvious CRD URLs 404.

On the sidecar cache question raised on the #10274 diff: with the VGS CRDs present all four caches populate and the controller reaches `controller initialized` normally. The hang is a symptom of the gate being on while the CRDs are missing, not something inherent.

API versions line up. v8.6.0 serves VGS `v1`, `v1beta1` and `v1beta2` with **v1beta2 as the storage version**, which is what Velero builds against (`pkg/backup/actions/csi/pvc_action.go:26`, used at `:821` and `:969`).

## A phase 4 prerequisite that isn't on the list yet

There is no `test/testdata/volume-group-snapshot-class/`. Velero selects a VolumeGroupSnapshotClass by `velero.io/csi-volumegroupsnapshot-class` (`pkg/apis/velero/v1/labels_annotations.go:199`) and csi-driver-host-path ships no VolumeGroupSnapshotClass at all, so a VGS spec cannot select one until that testdata exists. It is the exact analogue of what #10236 added for VolumeSnapshotClass. Happy to open it as a small first PR if you want phase 4 unblocked early.

Related, hostpath ships no StorageClass either, so any CSI e2e work has to supply one regardless of how #8765 lands.

## Phase 3 costing

I measured the current matrix so the CI budget question has numbers behind it. From run 33773942233, all green on `main`:

| Label set | Jobs | Mean | Max |
|---|---:|---:|---:|
| `ResourceFiltering && !FSBackup` | 13 | 18.9m | 20m |
| `ResourceModifier \|\| ...` | 13 | 16.1m | 16m |
| `NamespaceMapping ...` | 13 | 13.4m | 14m |
| `Basic && ...` | 13 | 6.6m | 7m |

52 jobs, about 715 minutes total, `timeout-minutes: 30` each. Note it is 13 k8s versions, not 12.

The CSI stack install measures at a mean of 204s over three cold runs, with a 20s spread, cluster creation excluded since the jobs already pay that.

The constraint that decides this is headroom rather than total cost. Folding the install into the existing four label sets adds 3.4m to every job, which leaves `ResourceFiltering` about 6.6m under its 30m timeout on every k8s version, permanently. That is a flake argument more than a cost one, and it also means a CSI problem can redden a currently green job, which is the #10274 failure mode again.

A fifth isolated label set looks better. Across all 13 versions it adds about 169m, or 24%. On a 3 version subset, oldest supported plus mid plus newest, it adds about 39m, or 5.5%. Given the work in #10458, #10390 and #9952 to shrink this matrix, the version subset seems like the lever that makes it affordable.

One honesty note on that: the per CSI job estimate of roughly 13m is extrapolated, not measured. It comes from the Basic job's real 6.6m, the measured 3.4m install, and a local 314s run of the two CSI specs. Local hardware is not a GitHub runner, so the only way to firm it up is to run a CSI job in CI, which is phase 3 itself.

## Questions

1. StorageClass strategy is the thing I'd most like settled before touching the workflow. Is #8765 the intended route, and is it available to pick up? It has been in backlog since 2025-03-07.
2. Scope: is KEP-3314 and snapshot metadata in, or is this VolumeSnapshot plus VGS green on kind? @PranjalManhgaye asked this on Aug 22 and it is still the main unknown for sizing phase 4.
3. PR decomposition for phase 3: installer, then StorageClass strategy, then the workflow job, then removing the now obsolete `CloudProvider == kind` skips. There are eight of those, in `backup/backup.go`, `backups/deletion.go`, `basic/namespace-mapping.go`, `migration/migration.go`, `upgrade/upgrade.go` and `bsl-mgmt/deletion.go`. Does that split match how you'd want to review it?
4. VGS spec shape: a new `Describe` block, or extend the `backup-volume-info` pattern the existing CSI specs use?

One piece of history that might be worth capturing in the issue, since it only exists in git. The exclusion was explicit before it was implicit: `e862b976a` ("Use labels instead of regex to filter E2E test cases", 2024-07-23) replaced `GINKGO_SKIP='SKIP_KIND|pv-backup|Restic|Snapshot|LongTime'` with the label matrix, which turned a legible negative filter into exclusion by omission and orphaned the `SKIP_KIND` label in the same stroke. That label is still set on three specs in `e2e_suite_test.go` and nothing filters on it, so those specs are simply never selected anywhere. Worth deciding whether the intent should be restored as a real filter or the label removed.

I also ran through #10267 on a fresh cluster and left notes there.

**Comment by krishhna24** [comment]
Phase 2 follow-up: I went through the `Snapshot`-labelled specs one at a time on kind to see which of the eight `CloudProvider == Kind` guards can actually come out.

Setup: kind v1.36.1, external-snapshotter v8.6.0, csi-driver-host-path, MinIO, Velero built from main. Guards commented out locally, both e2e StorageClasses pointed at `hostpath.csi.k8s.io`, and Kibishii's kind StorageClass patched the same way. All local, nothing committed. One spec per run.

| Spec | Result | Time |
| --- | --- | --- |
| `NamespaceMapping Single Snapshot` | pass | 537s |
| `NamespaceMapping Multiple Snapshot` | pass | 741s |
| `Basic Snapshot AdditionalBSL` | fail (0/2) | 457s |
| `Basic Snapshot RetainPV` | fail | 49s |
| `Backups Deletion Snapshot` | fail | 256s |
| `BSL Deletion Snapshot` | fail | 148s |
| `Upgrade Snapshot` (from v1.17.2) | fail | 237s |

Not run: `Migration` needs two clusters, `TTL` is `LongTime`, and `NativeSnapshot` can't work without a volume snapshotter plugin.

### What can be removed today

`basic/namespace-mapping.go:38-40,74-78`. Both specs back up with `--snapshot-volumes`, restore into the mapped namespace, and pass Kibishii's data verification.

Worth noting these two are not `Skip()` guards. `:38` zeroes the Kibishii data on kind and `:74` forces `--snapshot-volumes=false`, so on kind these specs have been running a quieter version of the test rather than being skipped. Removing the branches makes them exercise the real path.

### What needs a code change first

`backup/backup.go:77`, `backups/deletion.go:50,91`, `bsl-mgmt/deletion.go:61` and `upgrade/upgrade.go:70`. In each of these the backup reaches `Completed` and the volumeinfo lists the volumes, so Velero does the work. The failure is in verification:

```
exceed waiting for snapshot created in cloud: || UNEXPECTED || - Snapshots count is not as
expected after backup ...: Cloud provider kind is not valid
```

`getProvider` (`test/util/providers/common.go:75`) handles aws, vsphere, gcp and azure, so kind falls into `default:`. `CheckSnapshotsInProvider` (`:127`) exempts only `VanillaZFS`. Related, `BuildSnapshotCheckPointFromVolumeInfo` (`test/util/velero/velero_utils.go:1319`) sets `EnableCSI` only for AWS and Azure, so on kind the checkpoint comes out with `EnableCSI=false` even with `FEATURES=EnableCSI`.

For kind the snapshots are in the cluster rather than in a cloud, so verifying from the `VolumeSnapshotContent`s that volumeinfo already points at would cover it without touching the cloud paths. Happy to sketch that if the approach sounds right.

`migration/migration.go:84` I'd leave alone for now, since it needs two clusters and I can't assess it on a single kind cluster.

### Two gaps on the Kibishii side

Both in `vmware-tanzu-experiments/distributed-data-generator`, and thanks @shubham-pampattiwar for confirming that's where they belong:

1. `kubernetes/yaml/kind/kibishiiKINDStorageClass.yaml` uses `rancher.io/local-path`, so Kibishii's PVCs aren't snapshottable by CSI even when Velero's own StorageClass points at hostpath. Every run above needed this patched locally.
2. `overlays/sc-reclaim-policy/` has no kind variant (aws, aws-csi, azure, azure-csi, gcp, vanilla-zfs, vsphere only). That's why `Basic Snapshot RetainPV` fails after 49s, at kustomize, before Velero is involved at all.

So phase 3 turns out to need a harness change alongside the workflow and StorageClass work. I've picked up #8765 for the StorageClass part and asked @kaovilai about flag naming there. I'll sketch the Kibishii overlay change next for review before opening anything upstream.
